src/main/java/org/elasticsearch/common/lucene/Lucene.java

```diff
@@ -724,4 +726,36 @@ public void delete() {
             throw new UnsupportedOperationException("This IndexCommit does not support deletions");
         }
     }
+
+    /**
+     * Returns <code>true</code> iff the store contains an index that contains segments that were
+     * not upgraded to the lucene 4.x format.
+     */
+    public static boolean indexNeeds3xUpgrading(Directory directory) throws IOException {
+        final String si =  SegmentInfos.getLastCommitSegmentsFileName(directory);
+        if (si != null) {
+            try (IndexInput input = directory.openInput(si, IOContext.READONCE)) {
+                return input.readInt() != CodecUtil.CODEC_MAGIC; // check if it's a 4.x commit point
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Upgrades the segments metadata of the index to match a lucene 4.x index format. In particular it ensures that each
+     * segment has a .si file even if it was written with lucene 3.x
+     */
+    public static boolean upgradeLucene3xSegmentsMetadata(Directory directory) throws IOException {
+        if (indexNeeds3xUpgrading(directory)) {
+            try (final IndexWriter iw = new IndexWriter(directory, new IndexWriterConfig(Version.LATEST, Lucene.STANDARD_ANALYZER)
+                    .setMergePolicy(NoMergePolicy.INSTANCE)
+                    .setOpenMode(IndexWriterConfig.OpenMode.APPEND))) {
+                Map<String, String> commitData = iw.getCommitData();
+                iw.setCommitData(commitData);
```

Comment:
Maybe add a comment that you do this to trick IW into committing when it thinks nothing changed :)